src/SprykerEco/Zed/AvalaraTax/Business/Calculator/MultiShipmentCartItemAvalaraTaxCalculator.php

```diff
@@ -0,0 +1,154 @@
+        ItemTransfer $itemTransfer,
+        ArrayObject $avalaraLineItemTransfers,
+        array $zipCodeRegionNameMap
+    ): ?AvalaraTransactionLineTransfer {
+        foreach ($avalaraLineItemTransfers as $avalaraLineItemTransfer) {
+            if (!$itemTransfer->getAmountOrFail()->equals($avalaraLineItemTransfer->getQuantityOrFail())) {
+                continue;
+            }
+
+            $itemShipmentAddressZipCode = $itemTransfer->getShipmentOrFail()->getShippingAddressOrFail()->getZipCodeOrFail();
+            if (
+                !isset($zipCodeRegionNameMap[$itemShipmentAddressZipCode])
+                || $zipCodeRegionNameMap[$itemShipmentAddressZipCode] !== $this->extractRegionNameFromAvalaraTransactionLineTransfer($avalaraLineItemTransfer)
+            ) {
+                continue;
+            }
+
+            return $avalaraLineItemTransfer;
+        }
+
+        return null;
+    }
+
+    /**
+     * @param \Generated\Shared\Transfer\AvalaraTransactionTransfer $avalaraTransactionTransfer
+     *
+     * @return array
+     */
+    protected function getRegionZipCodeMap(AvalaraTransactionTransfer $avalaraTransactionTransfer): array
+    {
+        $zipCodeRegionMap = [];
+
+        $avalaraTransactionAddresses = $this->utilEncodingService->decodeJson($avalaraTransactionTransfer->getAddressesOrFail(), true);
+        foreach ($avalaraTransactionAddresses as $avalaraTransactionAddress) {
+            if (array_key_exists($avalaraTransactionAddress[static::KEY_POSTAL_CODE], $zipCodeRegionMap)) {
+                continue;
+            }
+
+            $zipCodeRegionMap[$avalaraTransactionAddress[static::KEY_POSTAL_CODE]] = $avalaraTransactionAddress[static::KEY_REGION];
+        }
+
+        return $zipCodeRegionMap;
+    }
+
+    /**
+     * @param \ArrayObject|\Generated\Shared\Transfer\AvalaraTransactionLineTransfer[] $avalaraTransactionLineTransfers
+     *
+     * @return \Generated\Shared\Transfer\AvalaraTransactionLineTransfer[][]
+     */
+    protected function getAvalaraTransactionLineTransfersIndexedByGroupKey(ArrayObject $avalaraTransactionLineTransfers): array
+    {
+        $indexedAvalaraTransactionLineTransfers = [];
+        foreach ($avalaraTransactionLineTransfers as $avalaraTransactionLineTransfer) {
+            $indexedAvalaraTransactionLineTransfers[$avalaraTransactionLineTransfer->getRef2OrFail()][] = $avalaraTransactionLineTransfer;
+        }
+
+        return $indexedAvalaraTransactionLineTransfers;
+    }
+
+    /**
+     * @param \Generated\Shared\Transfer\AvalaraTransactionLineTransfer $avalaraTransactionLineTransfer
+     *
+     * @return string|null
+     */
+    protected function extractRegionNameFromAvalaraTransactionLineTransfer(AvalaraTransactionLineTransfer $avalaraTransactionLineTransfer): ?string
+    {
+        $avalaraTransactionDetails = $this->utilEncodingService->decodeJson($avalaraTransactionLineTransfer->getDetailsOrFail(), true);
```

Comment:
* Try to decode it to `stdClass`, in this case it will be more convenient to refer to Avalara's models.

Reply:
Resolved.